js-miniapp-sample/src/pages/purchase.js

```diff
@@ -1,91 +1,309 @@
+  },
+  success: {
+    color: green[500],
+    marginTop: 20,
+  },
+  rootUserGroup: {
+    alignItems: 'center',
+  },
+  formInput: {
+    width: '90%',
+    marginTop: 10,
+  },
+  rootCardActions: {
     justifyContent: 'center',
+  },
+  caseSelector: {
+    marginTop: 5,
+  },
+  button: {
+    marginBottom: 15,
+  },
+  dataFormsWrapper: {
     display: 'flex',
     flexDirection: 'column',
-    alignItems: 'center',
-    fontSize: 18,
-    color: theme.color.primary,
-    fontWeight: 'bold',
-  },
-  actions: {
     justifyContent: 'center',
+    alignItems: 'center',
   },
-  textfield: {
-    width: '80%',
-    maxWidth: 300,
-    background: 'white',
-    '& input': {
-      color: theme.color.primary,
-      lineHeight: '1.5em',
-      fontSize: '1.2em',
+  paper: {
+    width: '100%',
+    paddingBottom: 10,
+    marginBottom: 20,
+    '&:last-child': {
+      marginBottom: 0,
     },
   },
+  red: {
+    color: red[500],
+  },
 }));
 
-function PurchaseComponent() {
+export const initialState = {
+  isLoading: false,
+  isError: false,
+};
+
+type State = {
+  isLoading: ?boolean,
+  isError: ?boolean,
+};
+
+type Action = {
+  type: string,
+};
+
+export const dataFetchReducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case 'FETCH_INIT':
+      return {
+        ...state,
+        isLoading: true,
+        isError: false,
+      };
+    case 'FETCH_SUCCESS':
+      return {
+        ...state,
+        isLoading: false,
+        isError: false,
+      };
+    case 'FETCH_FAILURE':
+      return {
+        ...initialState,
+        isLoading: false,
+        isError: true,
+      };
+
+    default:
+      throw Error('Unknown action type');
+  }
+};
+
+type PurchaseProductProps = {
+  purchasedProduct: PurchasedProduct,
+  purchaseProductUsing: (itemId: string) => Promise<PurchasedProduct>,
+};
+
+function PurchaseComponent(props: PurchaseProductProps) {
+  const [state, dispatch] = useReducer(dataFetchReducer, initialState);
   const classes = useStyles();
+
   let inputValue = '';
 
   const handleInput = (e: SyntheticInputEvent<HTMLInputElement>) => {
     e.preventDefault();
     inputValue = e.currentTarget.value;
   };
 
-  const purchaseItem = () => {
-    MiniApp.purchaseService
-      .purchaseItemWith(inputValue)
-      .then((success) => {
-        console.log('Success:', success);
-      })
-      .catch((error) => {
-        console.error(error);
+  const buttonClassname = clsx({
+    [classes.buttonFailure]: state.isError,
+    [classes.buttonSuccess]: !state.isError,
+  });
+
+  function BuyProduct() {
+    props
+      .purchaseProductUsing(inputValue)
+      .then(() => dispatch({ type: 'FETCH_SUCCESS' }))
+      .catch((e) => {
+        console.error(e);
+        dispatch({ type: 'FETCH_FAILURE' });
       });
-  };
+  }
 
-  return (
-    <GreyCard>
-      <CardContent className={classes.content}>
+  function handleClick(e) {
+    if (!state.isLoading) {
+      e.preventDefault();
```

Comment:
I don't think you need to use `preventDefault` for a React component Button.

Reply:
Done